### DISCUSSION TITLE
no-embed script feature not documented

### DISCUSSION THREAD

**Discussion by mikolasan** [ipxe/ipxe#441] [General]
I'm loading iPXE on UEFI firmware from USB drive following [this process](https://gist.github.com/AdrianKoshka/5b6f8b6803092d8b108cda2f8034539a), and I need to add a script. From [the docs](https://ipxe.org/scripting) I understand that my the only option is to embed it in, but I found in git history [this commit](https://github.com/ipxe/ipxe/commit/a3f1e8fb6707811e6eb90e339d7ebe813fd89a63) and in discussion #333 @mcb30 mentioned that 

> You may also want to consider using an `autoexec.ipxe` script loaded from the disk, so that you don't have to recompile iPXE each time.

Should some description also be added to the docs on ipxe.org?

**Comment by NiKiZe** [comment]
One consideration is that it only works in EFI mode, or rather from disk, will users try to put `autoexec.ipxe` on their TFTP expect it to work?

Below https://ipxe.org/scripting#embedded_scripts add "When running iPXE from USB drive, `autoexec.ipxe` can be used, it is detected automatically in EFI mode, for pcbios you need to ....."

**Comment by GethDeeo** [comment]
I'm too, in need of more information about this feature as I was unable to make it work.
iPXE (does not matter if downloaded from [boot.ipxe.org](https://boot.ipxe.org/ipxe.efi) or built one) boots, but my `autoexec.ipxe` containing only a simple `shell ||` line and placed in the root of the drive is seemingly ignored.

**Reply by mcb30** [reply]
It's an incomplete feature at present.  The most recent commit was https://github.com/ipxe/ipxe/commit/6ba671acd, which adds support for loading `autoexec.ipxe` via TFTP in a UEFI build.  Several variants (e.g. UEFI HTTP boot, or legacy PXE boot) are not yet implemented.

In your specific case (loading from disk): with a UEFI build it should work.  With a legacy BIOS build you will need to build your image manually using e.g. `make bin/ipxe.lkrn && ./util/genfsimg -s your_script.ipxe -o ipxe.img bin/ipxe.lkrn`.  You should then be able to edit the resulting `autoexec.ipxe` script within the root of the `ipxe.img` disk image.

**Reply by NiKiZe** [reply]
The script still needs to be a valid iPXE script.
That is first line starting with `#!ipxe` no BOM
So that makes it minimal 2 lines.

**Reply by GethDeeo** [reply]
Yep, that was it. I got overzealous while making a minimal working example :D

**Reply by GethDeeo** [reply]
Ok, so USB is working, TFTP is next (same binary, same script) :)
I have dnsmasq with the following:
```
dhcp-option=option:bootfile-name,ipxe.efi
dhcp-option=option:server-ip-address,192.168.1.12
```
iPXE gets downloaded (`tftp://...`), but it does not break the loop, just downloads iPXE again. Script is present and accessible at the same location as the binary, but I don't see any mention of autoexec in the tftp logs. Any hints? O_o

**Reply by mcb30** [reply]
@GethDeeo build with `DEBUG=efi_autoexec`

**Reply by GethDeeo** [reply]
Sorry for the pics, no idea how to capture text from hyper-v :O

First time:
![1](https://user-images.githubusercontent.com/21061348/150888485-34efdf33-caa9-410b-911f-a30585047cae.jpg)
Reentering "loop":
![2](https://user-images.githubusercontent.com/21061348/150888497-d430115c-15d0-4042-ab20-416b255881fa.jpg)

**Reply by tonho911** [reply]
![447530849_440685398901573_4048597028950067939_n](https://github.com/ipxe/ipxe/assets/148846306/24b8e221-263c-4bba-a81e-9b546d1d8389)
> @GethDeeo build with `DEBUG=efi_autoexec`

@mcb30 I got this error when trying to disable autoexec.ipxe with the command make bin-x86_64-efi/snponly.efi DEBUG=efi_autoexec

NEP file downloaded successfully.
IPXE initialising devices..
EFI Pciroot (0x0) /Pci (0x1, 0×2)/Pci (0x0, 0X2)/PCi (0X7,0X0) /Pci (0X0 , 0X0) /MAC (7085C26 44F5F0x0) /IPV4(0.0.0.0,0X0,DHCP,0.0.0.0,0.0.0.0,0.0.0.0) found no SimpleFileSys tem: Error 0x7f45e0Be (https://ipxe.org/7f45e08e)
EFI PciRoot (0X0) /Pci (0x1,0×2) /Pci (0X0,0×2)/Pci (0X7,0X0) /Pci (0X0, 0X0)/MAC (7085C26 44F5F ,0x0) /IPV4(0.0.0.0,0x0,DHCP,0.0.0.0,0.0.0.0,0.0.0.0) found ManagedNetworksb on PciRoot (0x0)/Pci (0x1,0×2) /Pci (0x0,0x2)/Pci (0X7,0X0)/Pci (0X0 , 0X0) /MAC (7085C26 44F5F ,0x0)
autoexec. ipxe. .. Not found (https: //ipxe.org/2d12618e)
EFI Pciroot (0x0) /Pci (0×1, 0×2) /Pci (0X0, 0X2) /Pci (0X7,0X0) /Pci (0X0, 0X0) /MAC (7085C26 44F5F,0x0) /IPV4 (0.0.0.0,0x0, DHCP,0.0.0.0,0.0.0.0,0.0.0.0) could not dounload aut Dexec. ipxe: Not found (https://ipxe.org/2d12618e)

is there a solution?

**Reply by NiKiZe** [reply]
> ![447530849_440685398901573_4048597028950067939_n](https://private-user-images.githubusercontent.com/148846306/337941701-24b8e221-263c-4bba-a81e-9b546d1d8389.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTc5MDk4ODgsIm5iZiI6MTcxNzkwOTU4OCwicGF0aCI6Ii8xNDg4NDYzMDYvMzM3OTQxNzAxLTI0YjhlMjIxLTI2M2MtNGJiYS1hODFlLTliNTQ2ZDFkODM4OS5qcGc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNjA5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDYwOVQwNTA2MjhaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT02ZGQ2YWY4ZjZkYWFjNmEzMmNiYzE2YzdjZDNlMTljMjc3YjY0Yzc2NjFlNDFmMzQ3MDUxMmI2MDMwMjQyOTRmJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.dE3BWExi9rF2GAeIm4djrVfYoGDjSAWZpBcGJ8Gg0x0)
> 
> > @GethDeeo build with `DEBUG=efi_autoexec`
> 
> @mcb30 I got this error when trying to disable autoexec.ipxe with the command make bin-x86_64-efi/snponly.efi DEBUG=efi_autoexec
> 
> NEP file downloaded successfully. IPXE initialising devices.. EFI Pciroot (0x0) /Pci (0x1, 0×2)/Pci (0x0, 0X2)/PCi (0X7,0X0) /Pci (0X0 , 0X0) /MAC (7085C26 44F5F0x0) /IPV4(0.0.0.0,0X0,DHCP,0.0.0.0,0.0.0.0,0.0.0.0) found no SimpleFileSys tem: Error 0x7f45e0Be (https://ipxe.org/7f45e08e) EFI PciRoot (0X0) /Pci (0x1,0×2) /Pci (0X0,0×2)/Pci (0X7,0X0) /Pci (0X0, 0X0)/MAC (7085C26 44F5F ,0x0) /IPV4(0.0.0.0,0x0,DHCP,0.0.0.0,0.0.0.0,0.0.0.0) found ManagedNetworksb on PciRoot (0x0)/Pci (0x1,0×2) /Pci (0x0,0x2)/Pci (0X7,0X0)/Pci (0X0 , 0X0) /MAC (7085C26 44F5F ,0x0) autoexec. ipxe. .. Not found (https: //ipxe.org/2d12618e) EFI Pciroot (0x0) /Pci (0×1, 0×2) /Pci (0X0, 0X2) /Pci (0X7,0X0) /Pci (0X0, 0X0) /MAC (7085C26 44F5F,0x0) /IPV4 (0.0.0.0,0x0, DHCP,0.0.0.0,0.0.0.0,0.0.0.0) could not dounload aut Dexec. ipxe: Not found (https://ipxe.org/2d12618e)
> 
> is there a solution?

Rather than error, it is information that no autoexec.ipxe was found, just ignore it if you don't want to use the feature.

**Reply by tonho911** [reply]
> > ![447530849_440685398901573_4048597028950067939_n](https://private-user-images.githubusercontent.com/148846306/337941701-24b8e221-263c-4bba-a81e-9b546d1d8389.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTc5MDk4ODgsIm5iZiI6MTcxNzkwOTU4OCwicGF0aCI6Ii8xNDg4NDYzMDYvMzM3OTQxNzAxLTI0YjhlMjIxLTI2M2MtNGJiYS1hODFlLTliNTQ2ZDFkODM4OS5qcGc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNjA5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDYwOVQwNTA2MjhaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT02ZGQ2YWY4ZjZkYWFjNmEzMmNiYzE2YzdjZDNlMTljMjc3YjY0Yzc2NjFlNDFmMzQ3MDUxMmI2MDMwMjQyOTRmJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.dE3BWExi9rF2GAeIm4djrVfYoGDjSAWZpBcGJ8Gg0x0)
> > > @GethDeeo build with `DEBUG=efi_autoexec`
> > 
> > 
> > @mcb30 I got this error when trying to disable autoexec.ipxe with the command make bin-x86_64-efi/snponly.efi DEBUG=efi_autoexec
> > NEP file downloaded successfully. IPXE initialising devices.. EFI Pciroot (0x0) /Pci (0x1, 0×2)/Pci (0x0, 0X2)/PCi (0X7,0X0) /Pci (0X0 , 0X0) /MAC (7085C26 44F5F0x0) /IPV4(0.0.0.0,0X0,DHCP,0.0.0.0,0.0.0.0,0.0.0.0) found no SimpleFileSys tem: Error 0x7f45e0Be (https://ipxe.org/7f45e08e) EFI PciRoot (0X0) /Pci (0x1,0×2) /Pci (0X0,0×2)/Pci (0X7,0X0) /Pci (0X0, 0X0)/MAC (7085C26 44F5F ,0x0) /IPV4(0.0.0.0,0x0,DHCP,0.0.0.0,0.0.0.0,0.0.0.0) found ManagedNetworksb on PciRoot (0x0)/Pci (0x1,0×2) /Pci (0x0,0x2)/Pci (0X7,0X0)/Pci (0X0 , 0X0) /MAC (7085C26 44F5F ,0x0) autoexec. ipxe. .. Not found (https: //ipxe.org/2d12618e) EFI Pciroot (0x0) /Pci (0×1, 0×2) /Pci (0X0, 0X2) /Pci (0X7,0X0) /Pci (0X0, 0X0) /MAC (7085C26 44F5F,0x0) /IPV4 (0.0.0.0,0x0, DHCP,0.0.0.0,0.0.0.0,0.0.0.0) could not dounload aut Dexec. ipxe: Not found (https://ipxe.org/2d12618e)
> > is there a solution?
> 
> Rather than error, it is information that no autoexec.ipxe was found, just ignore it if you don't want to use the feature.

Oh ic, that means the autoexec.ipxe feature can't be removed, right?
and if you can remove autoexec.ipxe, where do you have to edit it, bro?
